### PR TITLE
add invariant tests

### PR DESCRIPTION
**Pull request by skhomuti**

### CODE REVIEW

**Review by madlabman — DISMISSED** [review]

File: test/fork/invariant/Invariants.t.sol [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+                no.totalDepositedKeys - no.totalExitedKeys,
+                no.stuckValidatorsCount,
+                "assert deposited - exited >= stuck"
+            );
+
+            assertGe(
+                no.totalAddedKeys,
+                no.depositableValidatorsCount + no.totalWithdrawnKeys,
+                "assert added >= depositable + withdrawn"
+            );
```
Comment (dgusakov) [inline test/fork/invariant/Invariants.t.sol:61]: This one looks strange even though the relation is correct. What do we want to check here exactly?

File: test/fork/invariant/Invariants.t.sol [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+            totalDepositedValidators += no.totalDepositedKeys;
+            totalDepositableValidators += no.depositableValidatorsCount;
+        }
+
+        (
+            uint256 _totalExitedValidators,
+            uint256 _totalDepositedValidators,
+            uint256 _depositableValidatorsCount
+        ) = csm.getStakingModuleSummary();
+        assertTrue(totalExitedValidators <= _totalExitedValidators);
```
Comment (dgusakov) [inline test/fork/invariant/Invariants.t.sol:96]: Why not strict equality?

File: test/fork/invariant/Invariants.t.sol [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+        (uint128 head, uint128 length) = csm.depositQueue();
+        for (uint128 i = 0; i < length; i++) {
+            queueIndex = head + i;
+            Batch item = csm.depositQueueItem(queueIndex);
+            batchKeys[item.noId()] += item.keys();
+            assertLe(
+                no.enqueuedCount,
+                item.keys(),
+                "assert enqueued <= batch keys"
+            );
```
Comment (dgusakov) [inline test/fork/invariant/Invariants.t.sol:117]: What if there are multiple batches?

File: test/fork/invariant/Invariants.t.sol [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+        }
+
+        for (uint256 noId = 0; noId < noCount; noId++) {
+            no = csm.getNodeOperator(noId);
+            assertGe(
+                no.enqueuedCount,
+                batchKeys[noId],
+                "assert enqueued >= batch keys"
```
Comment (dgusakov) [inline test/fork/invariant/Invariants.t.sol:125]: This is opposite to the condition above. I suggest keeping only this one

File: test/fork/invariant/Invariants.t.sol (hunk 1 above)
Comment (skhomuti) [reply test/fork/invariant/Invariants.t.sol:61]: that withdrawn are considered correctly

File: test/fork/invariant/Invariants.t.sol (hunk 2 above)
Comment (skhomuti) [reply test/fork/invariant/Invariants.t.sol:96]: mybad

File: test/fork/invariant/Invariants.t.sol (hunk 3 above)
Comment (skhomuti) [reply test/fork/invariant/Invariants.t.sol:117]: removed due to it duplicates the check below

**Review by dgusakov — DISMISSED** [review]

File: test/fork/invariant/Invariants.t.sol [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+
+        uint256 noCount = csm.getNodeOperatorsCount();
+        NodeOperator memory no;
+        for (uint256 noId = 0; noId < noCount; noId++) {
+            no = csm.getNodeOperator(noId);
+
+            assertTrue(
+                no.totalAddedKeys <=
+                    csm.MAX_SIGNING_KEYS_PER_OPERATOR_BEFORE_PUBLIC_RELEASE()
+            );
```
Comment (madlabman) [inline test/fork/invariant/Invariants.t.sol:143]: assertGe?

**Review by dgusakov — APPROVED** [review]

**Review by madlabman — APPROVED** [review]

**Pull request merged**